File: flutter_app/lib/presentation/home/home_view_model.dart
import 'package:flutter/material.dart';
import 'package:firebase_auth/firebase_auth.dart' as auth;
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:uuid/uuid.dart';

import '../../domain/model/data_models.dart';

class HomeUiState {
  final bool isLoading;
  final User? userData;
  final List<User> topCaregivers;
  final Booking? activeBooking;
  final ElderProfile? activeElder;
  final String? errorMessage;
  final bool isBirthday;
  final List<ElderProfile> clientElders;

  HomeUiState({
    this.isLoading = true,
    this.userData,
    this.topCaregivers = const [],
    this.activeBooking,
    this.activeElder,
    this.errorMessage,
    this.isBirthday = false,
    this.clientElders = const [],
  });

  HomeUiState copyWith({
    bool? isLoading,
    User? userData,
    List<User>? topCaregivers,
    Booking? activeBooking,
    ElderProfile? activeElder,
    String? errorMessage,
    bool? isBirthday,
    List<ElderProfile>? clientElders,
    bool clearActiveData = false,
  }) {
    return HomeUiState(
      isLoading: isLoading ?? this.isLoading,
      userData: userData ?? this.userData,
      topCaregivers: topCaregivers ?? this.topCaregivers,
      activeBooking: clearActiveData ? null : (activeBooking ?? this.activeBooking),
      activeElder: clearActiveData ? null : (activeElder ?? this.activeElder),
      errorMessage: errorMessage ?? this.errorMessage,
      isBirthday: isBirthday ?? this.isBirthday,
      clientElders: clientElders ?? this.clientElders,
    );
  }
}

class HomeViewModel extends ChangeNotifier {
  final auth.FirebaseAuth _auth = auth.FirebaseAuth.instance;
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  HomeUiState _uiState = HomeUiState();
  HomeUiState get uiState => _uiState;

  List<Review> _reviewsState = [];
  List<Review> get reviewsState => _reviewsState;

  bool _isReviewsLoading = false;
  bool get isReviewsLoading => _isReviewsLoading;

  List<SubscriptionPlan> _subscriptionPlans = [];
  List<SubscriptionPlan> get subscriptionPlans => _subscriptionPlans;

  HomeViewModel() {
    fetchUserData();
    fetchSubscriptionPlans();
  }

  void _setUiState(HomeUiState state) {
    _uiState = state;
    notifyListeners();
  }

  void _setReviewsState(List<Review> reviews) {
    _reviewsState = reviews;
    notifyListeners();
  }

  void _setReviewsLoading(bool loading) {
    _isReviewsLoading = loading;
    notifyListeners();
  }

  void _setSubscriptionPlans(List<SubscriptionPlan> plans) {
    _subscriptionPlans = plans;
    notifyListeners();
  }

  void fetchSubscriptionPlans() {
    _firestore.collection('subscription_plans').snapshots().listen((snapshot) {
      final plans = snapshot.docs.map((doc) => SubscriptionPlan.fromMap(doc.data())).toList();
      _setSubscriptionPlans(plans);
    });
  }

  void fetchUserData() {
    _setUiState(_uiState.copyWith(isLoading: true));
    final userId = _auth.currentUser?.uid;
    if (userId == null) {
      _setUiState(_uiState.copyWith(isLoading: false, errorMessage: "Not logged in"));
      return;
    }

    _firestore.collection('users').doc(userId).snapshots().listen((document) async {
      if (document.exists) {
        final user = User.fromMap(document.data()!);

        // Auto-verify Caregivers/Nurses if KYC is APPROVED
        bool isCaregiverRole = user.role == "CAREGIVER" || user.role == "NURSE";
        if (isCaregiverRole && user.kycStatus == "APPROVED" && !user.isVerified) {
          await _firestore.collection('users').doc(userId).update({"isVerified": true});
        }

        bool isUserBirthday = false;
        if (user.dob.isNotEmpty) {
          try {
            final parts = user.dob.replaceAll("/", "-").split("-");
            if (parts.length >= 3) {
              int? month;
              int? day;
              if (parts[0].length == 4) {
                month = int.tryParse(parts[1]);
                day = int.tryParse(parts[2]);
              } else {
                day = int.tryParse(parts[0]);
                month = int.tryParse(parts[1]);
              }

              final now = DateTime.now();
              if (month == now.month && day == now.day) {
                isUserBirthday = true;
              }
            }
          } catch (_) {}
        }

        if (user.role == "CLIENT") {
          _firestore.collection("elders").where("clientId", isEqualTo: userId).snapshots().listen((snap) {
            final elders = snap.docs.map((doc) => ElderProfile.fromMap(doc.data())).toList();
            _setUiState(_uiState.copyWith(clientElders: elders));
          });

          try {
            final caregiversSnapshot = await _firestore.collection("users")
                .where("role", whereIn: ["CAREGIVER", "NURSE"])
                .get();

            final todayStr = "${DateTime.now().year}-${DateTime.now().month.toString().padLeft(2, '0')}-${DateTime.now().day.toString().padLeft(2, '0')}";
            
            var caregiversList = caregiversSnapshot.docs
                .map((doc) => User.fromMap(doc.data()))
                .where((it) => (it.isVerified || it.kycStatus == "APPROVED") && !it.isBusy && !it.onLeaveDates.contains(todayStr))
                .toList();

            caregiversList.sort((a, b) => b.rating.compareTo(a.rating));
            if (caregiversList.length > 5) caregiversList = caregiversList.sublist(0, 5);

            _setUiState(_uiState.copyWith(isLoading: false, userData: user, topCaregivers: caregiversList, isBirthday: isUserBirthday));
          } catch (e) {
            _setUiState(_uiState.copyWith(isLoading: false, userData: user, isBirthday: isUserBirthday));
          }
        } else if (user.role == "CAREGIVER" || user.role == "NURSE") {
          _firestore.collection("bookings")
              .where("caregiverId", isEqualTo: userId)
              .where("status", whereIn: ["ACCEPTED", "IN_PROGRESS"])
              .snapshots()
              .listen((snap) async {
            if (snap.docs.isNotEmpty) {
              final booking = Booking.fromMap(snap.docs.first.data());
              _setUiState(_uiState.copyWith(activeBooking: booking));
              if (booking.elderId.isNotEmpty) {
                final elderDoc = await _firestore.collection("elders").doc(booking.elderId).get();
                if (elderDoc.exists) {
                  _setUiState(_uiState.copyWith(activeElder: ElderProfile.fromMap(elderDoc.data()!)));
                }
              }
            } else {
              _setUiState(_uiState.copyWith(clearActiveData: true));
            }
          });
          _setUiState(_uiState.copyWith(isLoading: false, userData: user, isBirthday: isUserBirthday));
        }
      } else {
        _setUiState(_uiState.copyWith(isLoading: false, errorMessage: "User data not found"));
      }
    }, onError: (e) {
      _setUiState(_uiState.copyWith(isLoading: false, errorMessage: e.toString()));
    });
  }

  void dismissBirthdayPopup() {
    _setUiState(_uiState.copyWith(isBirthday: false));
  }

  Future<void> purchaseSubscriptionPlan(SubscriptionPlan plan, String elderId, Function(bool, String) onResult) async {
    try {
      final clientId = _auth.currentUser?.uid;
      if (clientId == null) return;
      final jobId = const Uuid().v4();

      final newBooking = Booking(
        id: jobId,
        clientId: clientId,
        caregiverId: '',
        elderId: elderId,
        status: "BROADCASTED",
        isEmergency: true,
        requestedTime: DateTime.now().millisecondsSinceEpoch,
        jobDescription: "Premium Package: ${plan.title}\n\n${plan.description}",
        isSubscriptionBooking: true,
        planName: plan.title,
        allowedCaregivers: plan.assignedCaregivers,
        totalAmount: plan.price,
        isPaid: true,
      );

      await _firestore.collection("bookings").doc(jobId).set(newBooking.toMap());

      final txId = const Uuid().v4();
      final txData = {
        "id": txId,
        "userId": clientId,
        "amount": plan.price,
        "type": "PACKAGE_FEE",
        "description": "Purchased Package: ${plan.title}",
        "timestamp": DateTime.now().millisecondsSinceEpoch,
      };
      await _firestore.collection("transactions").doc(txId).set(txData);

      onResult(true, "සාර්ථකව මිලදී ගන්නා ලදී! Admin අනුයුක්ත කළ සේවකයින්ට පණිවිඩය යවන ලදී.");
    } catch (e) {
      onResult(false, "දෝෂයකි: $e");
    }
  }

  Future<void> fetchReviewsForCaregiver(String caregiverId) async {
    _setReviewsLoading(true);
    try {
      final snapshot = await _firestore.collection("reviews").where("caregiverId", isEqualTo: caregiverId).get();
      final reviews = snapshot.docs.map((doc) => Review.fromMap(doc.data())).toList();
      reviews.sort((a, b) => b.timestamp.compareTo(a.timestamp));
      _setReviewsState(reviews);
    } catch (e) {
      _setReviewsState([]);
    } finally {
      _setReviewsLoading(false);
    }
  }

  void clearReviews() {
    _setReviewsState([]);
  }

  void logout() {
    _auth.signOut();
  }
}

File: app/src/main/java/com/kina/care/presentation/home/HomeViewModel.kt
package com.kina.care.presentation.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.kina.care.domain.model.Booking
import com.kina.care.domain.model.ElderProfile
import com.kina.care.domain.model.Review
import com.kina.care.domain.model.SubscriptionPlan
import com.kina.care.domain.model.User
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.UUID

data class HomeUiState(
    val isLoading: Boolean = true,
    val userData: User? = null,
    val topCaregivers: List<User> = emptyList(),
    val activeBooking: Booking? = null,
    val activeElder: ElderProfile? = null,
    val errorMessage: String? = null,
    val isBirthday: Boolean = false,
    val clientElders: List<ElderProfile> = emptyList()
)

class HomeViewModel : ViewModel() {
    private val auth: FirebaseAuth = FirebaseAuth.getInstance()
    private val firestore: FirebaseFirestore = FirebaseFirestore.getInstance()

    private val _uiState = MutableStateFlow(HomeUiState())
    val uiState: StateFlow<HomeUiState> = _uiState.asStateFlow()

    private val _reviewsState = MutableStateFlow<List<Review>>(emptyList())
    val reviewsState: StateFlow<List<Review>> = _reviewsState.asStateFlow()
    var isReviewsLoading = MutableStateFlow(false)

    private val _subscriptionPlans = MutableStateFlow<List<SubscriptionPlan>>(emptyList())
    val subscriptionPlans: StateFlow<List<SubscriptionPlan>> = _subscriptionPlans.asStateFlow()

    init {
        fetchUserData()
        fetchSubscriptionPlans()
    }

    private fun fetchSubscriptionPlans() {
        firestore.collection("subscription_plans")
            .addSnapshotListener { snapshot, error ->
                if (error != null) return@addSnapshotListener
                if (snapshot != null) {
                    _subscriptionPlans.value = snapshot.documents.mapNotNull { it.toObject(SubscriptionPlan::class.java) }
                }
            }
    }

    fun fetchUserData() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)
            try {
                val userId = auth.currentUser?.uid
                if (userId != null) {
                    // 🌟 Fix: get() වෙනුවට addSnapshotListener() — real-time update ලැබෙනවා
                    firestore.collection("users").document(userId)
                        .addSnapshotListener { document, _ ->
                            if (document != null && document.exists()) {
                                val user = document.toObject(User::class.java)

                                // 🌟 Fix: CAREGIVER/NURSE ලට මක් isVerified = true auto-fix
                                // CLIENT ලා registration fee ගෙව්වාට පස්සෙ verified වෙනවා
                                val isCaregiverRole = user?.role == "CAREGIVER" || user?.role == "NURSE"
                                if (user != null && isCaregiverRole && user.kycStatus == "APPROVED" && !user.isVerified) {
                                    firestore.collection("users").document(userId)
                                        .update("isVerified", true)
                                }

                                var isUserBirthday = false
                                if (user != null && user.dob.isNotEmpty()) {
                                    try {
                                        val cleanDob = user.dob.replace("/", "-")
                                        val parts = cleanDob.split("-")
                                        if (parts.size >= 3) {
                                            val month: Int?
                                            val day: Int?
                                            if (parts[0].length == 4) {
                                                month = parts[1].toIntOrNull()
                                                day = parts[2].toIntOrNull()
                                            } else {
                                                day = parts[0].toIntOrNull()
                                                month = parts[1].toIntOrNull()
                                            }

                                            val cal = Calendar.getInstance()
                                            if (month == cal.get(Calendar.MONTH) + 1 && day == cal.get(Calendar.DAY_OF_MONTH)) {
                                                isUserBirthday = true
                                            }
                                        }
                                    } catch (e: Exception) { e.printStackTrace() }
                                }

                                if (user?.role == "CLIENT") {
                                    firestore.collection("elders").whereEqualTo("clientId", userId).addSnapshotListener { snap, _ ->
                                        if (snap != null) {
                                            _uiState.value = _uiState.value.copy(clientElders = snap.documents.mapNotNull { it.toObject(ElderProfile::class.java) })
                                        }
                                    }

                                    viewModelScope.launch {
                                        try {
                                            val caregiversSnapshot = firestore.collection("users")
                                                .whereIn("role", listOf("CAREGIVER", "NURSE"))
                                                .get().await()

                                            val todayStr = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(Date())
                                            val caregiversList = caregiversSnapshot.documents.mapNotNull { it.toObject(User::class.java) }
                                                .filter { (it.isVerified || it.kycStatus == "APPROVED") && !it.isBusy && !it.onLeaveDates.contains(todayStr) }
                                                .sortedByDescending { it.rating }
                                                .take(5)

                                            _uiState.value = _uiState.value.copy(isLoading = false, userData = user, topCaregivers = caregiversList, isBirthday = isUserBirthday)
                                        } catch (e: Exception) {
                                            _uiState.value = _uiState.value.copy(isLoading = false, userData = user, isBirthday = isUserBirthday)
                                        }
                                    }
                                } else if (user?.role == "CAREGIVER" || user?.role == "NURSE") {
                                    firestore.collection("bookings")
                                        .whereEqualTo("caregiverId", userId)
                                        .whereIn("status", listOf("ACCEPTED", "IN_PROGRESS"))
                                        .addSnapshotListener { snap, error ->
                                            if (error != null) return@addSnapshotListener
                                            if (snap != null && !snap.isEmpty) {
                                                val booking = snap.documents[0].toObject(Booking::class.java)
                                                _uiState.value = _uiState.value.copy(activeBooking = booking)
                                                if (booking != null && booking.elderId.isNotEmpty()) {
                                                    firestore.collection("elders").document(booking.elderId).get()
                                                        .addOnSuccessListener { elderDoc ->
                                                            if (elderDoc.exists()) {
                                                                _uiState.value = _uiState.value.copy(activeElder = elderDoc.toObject(ElderProfile::class.java))
                                                            }
                                                        }
                                                }
                                            } else {
                                                _uiState.value = _uiState.value.copy(activeBooking = null, activeElder = null)
                                            }
                                        }
                                    _uiState.value = _uiState.value.copy(isLoading = false, userData = user, isBirthday = isUserBirthday)
                                }
                            } else {
                                _uiState.value = _uiState.value.copy(isLoading = false, errorMessage = "User data not found")
                            }
                        }
                }
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(isLoading = false, errorMessage = e.localizedMessage)
            }
        }
    }

    fun dismissBirthdayPopup() { _uiState.value = _uiState.value.copy(isBirthday = false) }

    fun purchaseSubscriptionPlan(plan: SubscriptionPlan, elderId: String, onResult: (Boolean, String) -> Unit) {
        viewModelScope.launch {
            try {
                val clientId = auth.currentUser?.uid ?: return@launch
                val jobId = UUID.randomUUID().toString()

                val newBooking = Booking(
                    id = jobId,
                    clientId = clientId,
                    caregiverId = null,
                    elderId = elderId,
                    status = "BROADCASTED",
                    isEmergency = true,
                    requestedTime = System.currentTimeMillis(),
                    jobDescription = "Premium Package: ${plan.title}\n\n${plan.description}",

                    isSubscriptionBooking = true,
                    planName = plan.title,
                    allowedCaregivers = plan.assignedCaregivers,

                    totalAmount = plan.price,
                    isPaid = true
                )

                firestore.collection("bookings").document(jobId).set(newBooking).await()

                val txId = UUID.randomUUID().toString()
                val txData = hashMapOf(
                    "id" to txId, "userId" to clientId, "amount" to plan.price,
                    "type" to "PACKAGE_FEE", "description" to "Purchased Package: ${plan.title}",
                    "timestamp" to System.currentTimeMillis()
                )
                firestore.collection("transactions").document(txId).set(txData).await()

                onResult(true, "සාර්ථකව මිලදී ගන්නා ලදී! Admin අනුයුක්ත කළ සේවකයින්ට පණිවිඩය යවන ලදී.")
            } catch (e: Exception) {
                onResult(false, "දෝෂයකි: ${e.message}")
            }
        }
    }

    fun fetchReviewsForCaregiver(caregiverId: String) {
        viewModelScope.launch {
            isReviewsLoading.value = true
            try {
                val snapshot = firestore.collection("reviews").whereEqualTo("caregiverId", caregiverId).get().await()
                _reviewsState.value = snapshot.documents.mapNotNull { it.toObject(Review::class.java) }.sortedByDescending { it.timestamp }
            } catch (e: Exception) {
                _reviewsState.value = emptyList()
            } finally {
                isReviewsLoading.value = false
            }
        }
    }

    fun clearReviews() { _reviewsState.value = emptyList() }

    fun logout() {
        auth.signOut()
    }
}